realm/src/main/java/io/realm/dynamic/DynamicRealmList.java

```diff
@@ -0,0 +1,139 @@
+/*
+ * Copyright 2015 Realm Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.realm.dynamic;
+
+import java.util.AbstractList;
+import java.util.List;
+
+import io.realm.Realm;
+import io.realm.internal.LinkView;
+
+/**
+ * {@link io.realm.RealmList} exposed using a dynamic API. All objects in the list must have the same schema even though
+ * they are accessed dynamically. Null values are not allowed in this list.
+ */
+public class DynamicRealmList extends AbstractList<DynamicRealmObject> {
+
+    private final LinkView linkView;
+    private final Realm realm;
+
+    DynamicRealmList(LinkView linkView, Realm realm) {
+        this.linkView = linkView;
+        this.realm = realm;
+    }
+
+    /**
+     * Adds the specified object at the end of this List.
+     *
+     * @param object the object to add.
+     * @return true
+     * @throws IllegalArgumentException if object is either {@code null} or has the wrong type.
+     */
+    @Override
+    public boolean add(DynamicRealmObject object) {
+        checkIsValidObject(object);
+        linkView.add(object.row.getIndex());
+        return true;
```

Comment:
A rare case could happen here. If the `size()` has been reached `Integer.MAX_VALUE`, `add` will succeed and return true. But user can never get the Object. Should it be more clear that just throw a IndexOutOfBoundsException here?